scipy/ndimage/src/ni_support.c

```diff
@@ -195,125 +195,105 @@ int NI_InitLineBuffer(PyArrayObject *array, int axis, npy_intp size1,
 }
 
 /* Extend a line in memory to implement boundary conditions: */
-int NI_ExtendLine(double *line, npy_intp length, npy_intp size1,
-                  npy_intp size2, NI_ExtendMode mode, double constant_value, char * errmsg)
+int NI_ExtendLine(double *buffer, npy_intp line_length,
+                  npy_intp size_before, npy_intp size_after,
+                  NI_ExtendMode extend_mode, double extend_value,
+                  char *errmsg)
 {
-    npy_intp ii, jj, length1, nextend, rextend;
-    double *l1, *l2, *l3, val;
-
-    switch (mode) {
-    case NI_EXTEND_WRAP:
-        /* deal with situation where data is shorter than needed
-           for filling the line */
-        nextend = size1 / length;
-        rextend = size1 - nextend * length;
-        l1 = line + size1 + length - rextend;
-        l2 = line;
-        for(ii = 0; ii < rextend; ii++)
-            *l2++ = *l1++;
-        for(ii = 0; ii < nextend; ii++) {
-            l1 = line + size1;
-            for(jj = 0; jj < length; jj++)
-                *l2++ = *l1++;
-        }
-        nextend = size2 / length;
-        rextend = size2 - nextend * length;
-        l1 = line + size1;
-        l2 = line + size1 + length;
-        for(ii = 0; ii < nextend; ii++) {
-            l3 = l1;
-            for(jj = 0; jj < length; jj++)
-                *l2++ = *l3++;
-        }
-        for(ii = 0; ii < rextend; ii++)
-            *l2++ = *l1++;
-        break;
-    case NI_EXTEND_MIRROR:
-        if (length == 1) {
-            l1 = line;
-            val = line[size1];
-            for(ii = 0; ii < size1; ii++)
-                *l1++ = val;
-            l1 = line + size1 + length;
-            val = line[size1 + length - 1];
-            for(ii = 0; ii < size2; ii++)
-                *l1++ = val;
-        } else {
-            length1 = length - 1;
-            nextend = size1 / length1;
-            rextend = size1 - nextend * length1;
-            l1 = line + size1 + 1;
-            l2 = l1 - 2;
-            for(ii = 0; ii < nextend; ii++) {
-                l3 = l1;
-                for(jj = 0; jj < length1; jj++)
-                    *l2-- = *l3++;
-                l1 -= length1;
+    double *first = buffer + size_before;
+    double *last = first + line_length;
+    double *src, *dst, val;
+
+    switch (extend_mode) {
+        /* aaaaaaaa|abcd|dddddddd */
+        case NI_EXTEND_NEAREST:
+            src = first;
+            dst = buffer;
+            val = *src;
+            while (size_before--) {
+                *dst++ = val;
```

Comment:
Hadn't thought of it, but after giving it some thought I think it would only complicate things:
 * for NEAREST and CONSTANT: we are copying the same value to many places, so `memcpy` won't work,
 * for REFLECT and MIRROR: the data has to be reversed, so `memcpy` won't work, and
 * for WRAP: we could use `memcpy` if we kept the approach of the current master, calculating how many times `line_length` fits into `size_before` (or `size_after`) and how many items are left over, then `memcpy` the full `line_length` anywhere from zero to several times, then `memcpy` the left over items once. I have serious doubts that this would be faster, since we would have to do a division, which is expensive. And it would certainly complicate the code, which I think is especially concise and elegant for WRAP.